### PR TITLE
implemented channel shuffle

### PR DESCRIPTION
**Pull request by abh2k**

### CODE REVIEW

**Review by zoq — COMMENTED** [review]
Your output is insane, thanks for looking into this.

File: src/mlpack/methods/ann/layer/channel_shuffle.hpp [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+#ifndef MLPACK_METHODS_ANN_LAYER_CHANNEL_SHUFFLE_HPP
+#define MLPACK_METHODS_ANN_LAYER_CHANNEL_SHUFFLE_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * Definition and Implementation of the Channel Shuffle Layer.
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/channel_shuffle.hpp:19]:
```suggestion
 * Definition and implementation of the Channel Shuffle Layer.
```

Minor spelling fix.

File: src/mlpack/methods/ann/layer/channel_shuffle.hpp [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+ *
+ * Channel Shuffle divide the channels/units in a tensor into groups
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/channel_shuffle.hpp:21]:
```suggestion
 * Channel Shuffle divides the channels/units in a tensor into groups
```

Minor spelling fix.

File: src/mlpack/methods/ann/layer/channel_shuffle_impl.hpp [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    const size_t groupCount):
+    inRowSize(inRowSize),
+    inColSize(inColSize),
+    depth(depth),
+    groupCount(groupCount),
+    batchSize(0)
+{
+  if (depth % groupCount != 0)
+  {
+    Log::Fatal << "Number of channels must be divisible by groupCount.!" << std::endl;
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/channel_shuffle_impl.hpp:49]:
```suggestion
    Log::Fatal << "Number of channels must be divisible by groupCount!" << std::endl;
```

File: src/mlpack/tests/ann_layer_test.cpp [hunk 4]
```diff
@@ -4656,6 +4656,50 @@ TEST_CASE("TransposedConvolutionWeightInitializationTest", "[ANNLayerTest]")
...
+                  << 3  << 19 << arma::endr
+                  << 4  << 20 << arma::endr
+                  << 13 << 9 << arma::endr
+                  << 14 << 10 << arma::endr
+                  << 15 << 11 << arma::endr
+                  << 16 << 12 << arma::endr
+                  << 5  << 21 << arma::endr
+                  << 6  << 22 << arma::endr
+                  << 7  << 23 << arma::endr
+                  << 8  << 24 << arma::endr;
```
Comment (zoq) [inline src/mlpack/tests/ann_layer_test.cpp:4691]: Do those values come from another framework?

Comment (abh2k) [reply src/mlpack/tests/ann_layer_test.cpp:4691]: I checked these values using Pytorch. I will link the implementation.

Comment (abh2k) [reply src/mlpack/tests/ann_layer_test.cpp:4691]: https://colab.research.google.com/drive/16ZATOft3wQ9u_IIeyNbupV-ouzuYHK9Z?usp=sharing

**Comment by abh2k** [comment]
@zoq  please review it. I have attached a colab file which was used for test case.

File: src/mlpack/methods/ann/layer/channel_shuffle.hpp [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+#ifndef MLPACK_METHODS_ANN_LAYER_CHANNEL_SHUFFLE_HPP
+#define MLPACK_METHODS_ANN_LAYER_CHANNEL_SHUFFLE_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * Definition and implementation of the Channel Shuffle Layer.
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/channel_shuffle.hpp:19]: I guess it makes sense to point to the ShuffleNet paper here for more information.

File: src/mlpack/methods/ann/layer/channel_shuffle_impl.hpp (hunk 3 above)
Comment (zoq) [reply src/mlpack/methods/ann/layer/channel_shuffle_impl.hpp:49]: Looks like this is still open?

File: src/mlpack/methods/ann/layer/channel_shuffle_impl.hpp [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+  {
+    assert(output.n_rows == inRowSize * inColSize * depth);
+    assert(output.n_cols == batchSize);
+  }
+
+
+  arma::cube inputAsCube(const_cast<arma::Mat<eT>&>(input).memptr(),
+      inRowSize, inColSize, depth * batchSize, false, false);
+  arma::cube outputAsCube(output.memptr(), inRowSize, inColSize,
+                          depth * batchSize, false, true);
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/channel_shuffle_impl.hpp:72]:
```suggestion
  arma::cube outputAsCube(output.memptr(), inRowSize, inColSize,
      depth * batchSize, false, true);
```

Just to use the same style as the line right above.

File: src/mlpack/methods/ann/layer/channel_shuffle_impl.hpp [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+  else
+  {
+    assert(output.n_rows == inRowSize * inColSize * depth);
+    assert(output.n_cols == batchSize);
+  }
+
+  arma::cube gradientAsCube(((arma::Mat<eT>&) gradient).memptr(), inColSize,
+      inColSize, depth * batchSize, false, false);
+  arma::cube outputAsCube(output.memptr(), inRowSize, inColSize,
+                          depth * batchSize, false, true);
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/channel_shuffle_impl.hpp:108]:
```suggestion
  arma::cube outputAsCube(output.memptr(), inRowSize, inColSize,
      depth * batchSize, false, true);
```

File: src/mlpack/tests/ann_layer_test.cpp (hunk 4 above)
Comment (zoq) [reply src/mlpack/tests/ann_layer_test.cpp:4691]: Can you add a comment right above the `outputExpected1` and note that the values where calculated using PyTorch.

File: src/mlpack/methods/ann/layer/channel_shuffle_impl.hpp [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+  batchSize = input.n_cols;
+
+  if (output.is_empty())
+    output.set_size(inRowSize * inColSize * depth, batchSize);
+  else
+  {
+    assert(output.n_rows == inRowSize * inColSize * depth);
+    assert(output.n_cols == batchSize);
+  }
+
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/channel_shuffle_impl.hpp:67]:
```suggestion
```

Remove the extra line here.

**Review by zoq — APPROVED** [review]
Looks great to me, just left a really minor last comment about the style.

File: src/mlpack/methods/ann/layer/channel_shuffle_impl.hpp [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+    for (size_t i = 0; i < groupSize; ++i)
+    {
+      for (size_t g = 0; g < groupCount; ++g, ++outChannelIdx)
+      {
+        size_t inChannelIdx = k * batchSize + g * groupSize + i;
+        outputAsCube.slice(outChannelIdx) = inputAsCube.slice(inChannelIdx);
+      }
+    }
+  }
+
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/layer/channel_shuffle_impl.hpp:86]:
```suggestion
```

**Review by kartikdutt18 — APPROVED** [review]
LGTM!
Just a minor comment. Read your mail, really nice work!

**Comment by zoq** [comment]
Really nice, thanks!

**Pull request merged**